Fix/20921 - Throw error for invalid character in flat rate cost.

TimBHowe opened the pull request:
Added a call to the evaluate_cost function in the flat rate shipping cost input's sanitize_cost function to throw an exception if there is an illegal character in the input in the back end to avoid the error going unnoticed until a user/customer hits it in the front-end

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:
This change would run the the evaluate_cost function on the flat rate shipping's input sanitize_cost function throwing any error in the back-end before they happen on the front end.
Closes #20921

### How to test the changes in this Pull Request:

1. In WooCommerce General Settings set the site currency to EUR €.
2. Set up a shipping zone with a flat rate shipping method.
3. Enter a USD $ currency symbol or some other invalid character like an under score _ into the cost field.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix: Added validation to the cost field for flat rate shipping.

TimBHowe committed:
Added a call to the evaluate_cost function in the sanitize_cost function to throw an exception if there is an illegal character in the input.

peterfabian commented:
Thanks for your contribution, @TimBHowe . Would you mind fixing the other input in Setup Wizard as per comment https://github.com/woocommerce/woocommerce/issues/20921#issuecomment-559537243?

TimBHowe committed:
Updated the wc_setup_shipping_save() function to sanitize the flatrate cost input

TimBHowe committed:
Added pattern attribute to the flat rate cost input field to only allow US and EU currency formating

TimBHowe committed:
Updated if statment to allow 0 values

TimBHowe commented:
Hey so I added some additional check to sanitize the input removing the set currency symbol when saving the value in the set up wiz.

Also added the pattern attribute which should support entering only numbers, periods and commas in both the US and EU currency format standards.

I wanted to also add the pattern attribute to the flat rate settings inputs, but as they are sent via ajax it would need to use something like the [checkValidity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/checkValidity) and didn't know if it would be ok to use/add.

Otherwise this should fix most of the problem. The sanitation on the wiz is not as strong and it might help to add additional instructions laying out the rules, but this should be a good start.

ObliviousHarmony commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -937,6 +937,7 @@ protected function get_wizard_shipping_methods( $country_code, $currency_code )
 						'default_value' => __( 'Cost', 'woocommerce' ),
 						'description'   => __( 'What would you like to charge for flat rate shipping?', 'woocommerce' ),
 						'required'      => true,
+						'pattern'       => '^[+-]?[0-9]{1,3}(?:[0-9]*(?:[.,][0-9]{2})?|(?:,[0-9]{3})*(?:\.[0-9]{2})?|(?:\.[0-9]{3})*(?:,[0-9]{2})?)$',
```
Due to the way browsers display pattern errors, I don't think we should add this constraint. Having a server-side error thrown for invalid formats should be sufficient.

ObliviousHarmony replied:
Hey @TimBHowe,

Did you want to wrap this PR up?

TimBHowe replied:
Ok, I'll remove it.

ObliviousHarmony commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -1279,13 +1281,19 @@ public function wc_setup_shipping_save() {
+
+				// Sanitize the cost field.
+				$domestic_cost = wc_clean( wp_unslash( $_POST['shipping_zones']['domestic'][ $domestic_method ] ) );
+				$domestic_cost = str_replace( array( get_woocommerce_currency_symbol(), html_entity_decode( get_woocommerce_currency_symbol() ) ), '', $domestic_cost );
```
Since the primary reason for this PR is to change the sanitization on the flat rate shipping method, I don't think we should make a change to the wizard that would affect others. I agree that the sanitization here could be a lot better though!

ObliviousHarmony replied:
This will work fine.

ObliviousHarmony commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -1297,13 +1305,19 @@ public function wc_setup_shipping_save() {
+
+				// Sanitize the cost field.
+				$intl_cost = wc_clean( wp_unslash( $_POST['shipping_zones']['intl'][ $intl_method ] ) );
+				$intl_cost = str_replace( array( get_woocommerce_currency_symbol(), html_entity_decode( get_woocommerce_currency_symbol() ) ), '', $intl_cost );
```
Same as the other!

ObliviousHarmony replied:
This will work fine.

ObliviousHarmony reviewed (changes requested):
Thank you very much for the contribution! I've left some comments for you!

TimBHowe committed:
remove the pattern constraint on the flatrate input in the setup wiz

ObliviousHarmony reviewed (approved):
Looks good to me, thanks for the fine work!
